src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -439,15 +673,23 @@ public void stop() throws InterruptedException {
 
   private void onStoragePut(Digest digest) {
     try {
-      backplane.addBlobLocation(digest, config.getPublicName());
+
+      // if the worker is a CAS member, it can send/modify blobs in the backplane.
+      if (isCasShard) {
+        backplane.addBlobLocation(digest, config.getPublicName());
+      }
     } catch (IOException e) {
       throw Status.fromThrowable(e).asRuntimeException();
     }
   }
 
   private void onStorageExpire(Iterable<Digest> digests) {
     try {
-      backplane.removeBlobsLocation(digests, config.getPublicName());
+
+      // if the worker is a CAS member, it can send/modify blobs in the backplane.
+      if (isCasShard) {
+        backplane.removeBlobsLocation(digests, config.getPublicName());
+      }
     } catch (IOException e) {
       throw Status.fromThrowable(e).asRuntimeException();
     }
```

Comment:
nit: if around try. even better for either not registering this callback for !isCasShard, or another method which has the exception translation.